subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/AbstractCompiler.java

```diff
@@ -34,18 +35,27 @@
     private final CommandLineToolContext invocationContext;
     private final boolean useCommandFile;
     private final BuildOperationExecutor buildOperationExecutor;
+    private final WorkerLeaseService workerLeaseService;
 
-    protected AbstractCompiler(BuildOperationExecutor buildOperationExecutor, CommandLineToolInvocationWorker commandLineToolInvocationWorker, CommandLineToolContext invocationContext, ArgsTransformer<T> argsTransformer, boolean useCommandFile) {
+    protected AbstractCompiler(BuildOperationExecutor buildOperationExecutor, CommandLineToolInvocationWorker commandLineToolInvocationWorker, CommandLineToolContext invocationContext, ArgsTransformer<T> argsTransformer, boolean useCommandFile, WorkerLeaseService workerLeaseService) {
         this.buildOperationExecutor = buildOperationExecutor;
         this.argsTransformer = argsTransformer;
         this.invocationContext = invocationContext;
         this.useCommandFile = useCommandFile;
         this.commandLineToolInvocationWorker = commandLineToolInvocationWorker;
+        this.workerLeaseService = workerLeaseService;
     }
 
     @Override
-    public WorkResult execute(T spec) {
-        buildOperationExecutor.runAll(commandLineToolInvocationWorker, newInvocationAction(spec));
+    public WorkResult execute(final T spec) {
+        final Action<BuildOperationQueue<CommandLineToolInvocation>> invocationAction = newInvocationAction(spec);
+
+        workerLeaseService.withoutProjectLock(new Runnable() {
+            @Override
+            public void run() {
+                buildOperationExecutor.runAll(commandLineToolInvocationWorker, invocationAction);
```

Comment:
I may be getting confused or misunderstanding, but I _think_ we are ok here.  By "common args" I assume you mean the args calculated by `AbstractCompiler.getArguments()` which fire the user-specified callbacks.  `getArguments()` is called as the very first step in `newInvocationAction()` which returns an `Action` object that uses _the results_ from calling `getArguments()` - it doesn't call `getArguments()` itself.  So the callbacks are executed _before_ the project lock is released.  Or am I misunderstanding your comment?